README.md
# PHPValidation

It is an open source library that can be used to test and validate data.

[![Version](http://poser.pugx.org/muhametsafak/PHPValidation/version)](https://packagist.org/packages/muhametsafak/PHPValidation)
[![License](http://poser.pugx.org/muhametsafak/PHPValidation/license)](https://packagist.org/packages/muhametsafak/PHPValidation)
[![Total Downloads](http://poser.pugx.org/muhametsafak/PHPValidation/downloads)](https://packagist.org/packages/muhametsafak/PHPValidation)

```
 ,ggggggggggg,   ,ggg,        gg  ,ggggggggggg,   ,ggg,         ,gg                                                                                      
dP"""88""""""Y8,dP""Y8b       88 dP"""88""""""Y8,dP""Y8a       ,8P          ,dPYb,               8I                I8                                    
Yb,  88      `8bYb, `88       88 Yb,  88      `8bYb, `88       d8'          IP'`Yb               8I                I8                                    
 `"  88      ,8P `"  88       88  `"  88      ,8P `"  88       88           I8  8I  gg           8I             88888888  gg                             
     88aaaad8P"      88aaaaaaa88      88aaaad8P"      88       88           I8  8'  ""           8I                I8     ""                             
     88"""""         88"""""""88      88"""""         I8       8I  ,gggg,gg I8 dP   gg     ,gggg,8I    ,gggg,gg    I8     gg     ,ggggg,     ,ggg,,ggg,  
     88              88       88      88              `8,     ,8' dP"  "Y8I I8dP    88    dP"  "Y8I   dP"  "Y8I    I8     88    dP"  "Y8ggg ,8" "8P" "8, 
     88              88       88      88               Y8,   ,8P i8'    ,8I I8P     88   i8'    ,8I  i8'    ,8I   ,I8,    88   i8'    ,8I   I8   8I   8I 
     88              88       Y8,     88                Yb,_,dP ,d8,   ,d8b,d8b,_ _,88,_,d8,   ,d8b,,d8,   ,d8b, ,d88b, _,88,_,d8,   ,d8'  ,dP   8I   Yb,
     88              88       `Y8     88                 "Y8P"  P"Y8888P"`Y8P'"Y888P""Y8P"Y8888P"`Y8P"Y8888P"`Y888P""Y888P""Y8P"Y8888P"    8P'   8I   `Y8
```

## Installation

```
composer require muhametsafak/phpvalidation
```

# Using

Create an object with the `\PHPValidation\PHPValidation` class.

```php
$val = new \PHPValidation\PHPValidation();
```

# Validation Rules

- `integer` : Verifies that the data is an integer.
- `float` : Verifies that the data is a floating point number.
- `numeric` : Verifies that the data is a numeric value.
- `string` : Verifies that the data is of a string type.
- `boolean` : Verifies that the data has a logical value or equivalent.
- `array` : Verifies that the data is an array.
- `mail` : Verifies that the data is an E-Mail address.
- `mailHost` : Verifies that the data is the E-Mail address using the specified host.
- `url` : Verifies that the data is a URL address.
- `urlHost` : Verifies that the data is a URL of the specified host (or subdomain).
- `empty` : Verifies that the data is empty.
- `required` : Verifies that the data is not null.
- `min` : Defines the minimum value the data can have. Specifies the minimum number of elements/characters if the data is a string or array.
- `max` : Defines the maximum value the data can have. Specifies the maximum number of elements/characters if the data is a string or array.
- `length` : Verifies that the number of characters is within the specified range.
- `regex` : Validates data with a predefined or postdefined regular expression.
- `date` : Attempts to verify that the data is a date.
- `dateFormat` : Verifies that the data is a date in a specified format.
- `ip` : Verifies that the data is IP address.
- `ipv4` : Verifies that the data is IPv4.
- `ipv6` : Verifies that the data is IPv6.
- `repeat` : Verifies that the data is the same as the value of a data key.
- `equals` : Verifies that the data is equivalent to the specified value.
- `startWith` : Verifies that the data starts with the specified value.
- `endWith` : Array or String. Verifies that the data ends with the specified value.
- `in` : Array or String. Verifies that the specified value is contained in the data.
- `notIn` : Array or String. Verifies that the specified value does not exist in the data.
- `alpha` : Verifies that the data is an alpha value.
- `alphaNum` : Verifies that the data is an alphanumeric value.
- `creditCard` : Verifies that the data is a credit card number.
- `only` : It validates only one of the specified values, case-insensitively.
- `strictOnly` : Case sensitive only validates that it is one of the values specified.
- `optional` : If there is data, it must obey the rules, but if there is no data with the corresponding key, the validation will not fail.

# Methods

### `version()`

Returns the version of the library.

```php
public function version(): string
```

### `withData()`

Returns a copy of the library, loading the specified data directory.

```php
public function withData(array $data = []): self
```

### `locale()`

Imports translations from the localization file.

```php
public function locale(?string $localePath = null): self
```

### `pattern()`

Defines a named regular expression pattern.

```php
public function pattern(string $name, string $pattern): self
```

### `rule()`

Adds a rule.

```php
public function rule(string|array $rule, string|array $dataKey): self
```

`$rule` : You can define multiple rules at once as a array or as a string separated by `|`. You can see the ready-made rules you can use [here](#validation-rules).

`$dataKey` : Defines the keys of the data to which the defined rules will be applied. If you want to add more than one, you can specify it as an array or as a string separated by `|`.

### `ruleClosure()`

It allows you to write your own validation method with a callable function.

```php
public function ruleClosure(\Closure $closure, array|string $dataKey, string $errorMsg = ''): self
```

### `ruleFunction()`

Defines a rule using a predefined function.

```php
public function ruleFunction(string $function, string|array $dataKey, string $errorMsg = ''): self
```

### `ruleMethod()`

Defines a class or object's method and rule.

```php
public function ruleMethod(string|object $classOrObject, string $method, string|array $dataKey, string $errorMsg = ''): self
```

### `validation()`

Validates data within specified rules.

```php
public function validation(): bool
```

### `data()`

Returns the value of a data.

```php
public function data(string $key): mixed|null
```

### `error()`

It adds an external error so that the validation fails.

```php
public function error(string $error, array $context = []): self
```

### `errors()`

Returns the errors that occurred as an array. An empty array is returned if there are no errors.

```php
public function errors(): array
```

### `labels()`

Defines the string to replace the key ({field}) of the data on errors.

```php
public function labels(array $labels): self
```

# Localization

Copy the code below and place it in a php file. After localizing and saving it, it tells the `locale()` method the full path of your file.

```php
<?php
return [
    'labels'            => [],
    'integer'           => '{field} must be an integer.',
    'float'             => '{field} must be an float.',
    'numeric'           => '{field} must be an numeric.',
    'string'            => '{field} must be an string.',
    'boolean'           => '{field} must be an boolean',
    'array'             => '{field} must be an Array.',
    'mail'              => '{field} must be an E-Mail address.',
    'mailHost'          => '{field} the email must be a {2} mail.',
    'url'               => '{field} must be an URL address.',
    'urlHost'           => 'The host of the {field} url must be {2}.',
    'empty'             => '{field} must be empty.',
    'required'          => '{field} cannot be left blank.',
    'min'               => '{field} must be greater than or equal to {2}.',
    'max'               => '{field} must be no more than {2}.',
    'length'            => 'The {field} length range must be {2}.',
    'regex'             => '{field} must match the {2} pattern.',
    'date'              => '{field} must be a date.',
    'dateFormat'        => '{field} must be a correct date format.',
    'ip'                => '{field} must be the IP Address.',
    'ipv4'              => '{field} must be the IPv4 Address.',
    'ipv6'              => '{field} must be the IPv6 Address.',
    'repeat'            => '{field} must be the same as {field1}',
    'equals'            => '{field} can only be {2}.',
    'startWith'         => '{field} must start with "{2}".',
    'endWith'           => '{field} must end with "{2}".',
    'in'                => '{field} must contain {2}.',
    'notIn'             => '{field} must not contain {2}.',
    'alpha'             => '{field} must contain only alpha characters.',
    'alphaNum'          => '{field} can only be alphanumeric.',
    'creditCard'        => '{field} must be a credit card number.',
    'only'              => 'The {field} value is not valid.',
    'strictOnly'        => 'The {field} value is not valid.',
    'custom_closure'    => '{field} could not be verified.',
    'custom_function'   => '{field} could not be verified.',
    'custom_method'     => '{field} could not be verified.',
];
```

```php
$validation->locale(__DIR__ . '/lang.php');
```

# Usage

```php
$validation = new \PHPValidation\PHPValidation($_POST ?? []);

/**
 * Both "username" and "password" must be 
 * a non-empty string.
 */
$validation->rule(['required', 'string'], ['username', 'password']);

/**
 * The method may need one or more parameters 
 * for the validation process.
 * 
 * You can send the required parameter 
 * in parenthesis "()".
 */
/**
 * In the examples below you can see an example 
 * of sending parameters for length validation.
 * 
 * "username" can be a minimum of 3 characters 
 * and a maximum of 255 characters.
 * "password" must be at least 8 characters.
 */
$validation->rule('length(3-255)', 'username');
$validation->rule('length(8-)', 'password');

/**
 * If the rule takes more than one parameter, 
 * you can send the parameters with comma ",".
 */
/**
 * Below you can see an example of sending multiple
 * parameters to the only() role, 
 * which may need multiple parameters to work.
 */
/**
 * Specified that if there is a value 
 * with the key "gender" it can 
 * only be "male" or "female".
 */
$validation->rule('only(male,female)|optional', 'gender');

if($validation->validation()){
    // Success
}else{
    // Error: Validation Failed.
    foreach($validation->errors() as $err){
        echo $err . '<br />' . \PHP_EOL;
    }
}
```

You can report an error with the `error()` method to fail the validation for any reason.

```php
$validation = new \PHPValidation\PHPValidation();

$test = $validation->withData([
    'username'  => 'muhametsafak',
    'password'  => 'Rd:3SvS?',
]);

$test->rule('required|string', ['username', 'password']);

// This will always fail.
$test->error('An error occurred during operations.');

if($test->validation()){
    // Success
}else{
    // Error : Failed.
}

```


# RULE Method Using

The Rule method allows a data to be tested using one or more Validation class methods.

The data to be tested is kept in an associative array. The `withData()` method is used to define this array.

With the `rule()` method, it is specified which method will be applied to which data.

The `validation()` method performs the tests specified before it by the `rule()` method. It is "true" if all results are "true". It is "false" if one or more of the operations is "false".

**Example 1:**

```php
$test = $val->withData(["dataName" => "info@test.com"]);
$test->rule("required", "dataName");
$test->rule("is_string", "dataName");
$test->rule("is_mail", "dataName");
if($test->validation()){
    //Success
}
```

## Custom Rule

If the internal rules of the class are not enough for you; You can set custom rules with `ruleClosure()`, `ruleFunction()`, `ruleMethod()` functions.

### Custom Closure Rule

```php
$test = $validation->withData(['input' => 'Data Content']);

$test->ruleClosure(function($data){
    if($data !== 'Text Val'){
        return false;
    }
}, 'input', 'Validation failed.');

if($test->validation()){
    // success
}
```

### Custom Function Rule

```php
function testVal($data)
{
    if(empty($data)){
        return false;
    }
}

$test = $validation->withData(['input' => 'Data Content']);

$test->ruleFunction('testVal', 'input', 'Validation failed.');

if($test->validation()){
    // success
}
```

### Custom Method Rule

It provides validation with a certain method of a class or object.

```php
class MyValidation
{
    public function testVal($data)
    {
        if($data != 'data content'){
            return false
        }
        return true;
    }
}

$test = $validation->withData(['input' => 'Data Content']);

$test->ruleMethod(MyValidation::class, 'testVal', 'input', 'Validation failed.');

if($test->validation()){
    // success
}
```

# Validation with Regular Expressions (REGEX)

You can test using regular expression patterns that you define with the `pattern()` method. The `regex()` method tests with the name of a pattern you define.

```php
$test = $val->withData(["dataName" => "Muhammet"]);
$test->pattern("patternName", "[a-zA-Z0-9]"); //Adding Pattern
$test->rule("regex(patternName)", "dataName");
if($test->validation()){
    //Success
}
```

***

## License

This library is written by [Muhammet ŞAFAK](https://www.muhammetsafak.com.tr) and distributed under the [GNU GPL 3.0](http://www.gnu.org/licenses/gpl-3.0.txt) license.

Please let me know if you encounter an error.

E-Mail : <info@muhammetsafak.com.tr> 


CHANGELOG.md
# Changelog

### 1.1 [2022.01.02]

- Added 3 methods (`ruleClosure()`, `ruleFunction()`, `ruleMethod()`) where you can write your own custom rules through external classes or functions.
- Added `locale()` method for you to create and import your own language files.
- Language switches are named more meaningfully. And the error messages have been edited in a more meaningful way.

### 1.0 [2021.10.03]

- First version

src/lang/en.php
<?php
return [
    'labels'            => [],
    'integer'           => '{field} must be an integer.',
    'float'             => '{field} must be an float.',
    'numeric'           => '{field} must be an numeric.',
    'string'            => '{field} must be an string.',
    'boolean'           => '{field} must be an boolean',
    'array'             => '{field} must be an Array.',
    'mail'              => '{field} must be an E-Mail address.',
    'mailHost'          => '{field} the email must be a {2} mail.',
    'url'               => '{field} must be an URL address.',
    'urlHost'           => 'The host of the {field} url must be {2}.',
    'empty'             => '{field} must be empty.',
    'required'          => '{field} cannot be left blank.',
    'min'               => '{field} must be greater than or equal to {2}.',
    'max'               => '{field} must be no more than {2}.',
    'length'            => 'The {field} length range must be {2}.',
    'regex'             => '{field} must match the {2} pattern.',
    'date'              => '{field} must be a date.',
    'dateFormat'        => '{field} must be a correct date format.',
    'ip'                => '{field} must be the IP Address.',
    'ipv4'              => '{field} must be the IPv4 Address.',
    'ipv6'              => '{field} must be the IPv6 Address.',
    'repeat'            => '{field} must be the same as {field1}',
    'equals'            => '{field} can only be {2}.',
    'startWith'         => '{field} must start with "{2}".',
    'endWith'           => '{field} must end with "{2}".',
    'in'                => '{field} must contain {2}.',
    'notIn'             => '{field} must not contain {2}.',
    'alpha'             => '{field} must contain only alpha characters.',
    'alphaNum'          => '{field} can only be alphanumeric.',
    'creditCard'        => '{field} must be a credit card number.',
    'only'              => 'The {field} value is not valid.',
    'strictOnly'        => 'The {field} value is not valid.',
    'custom_closure'    => '{field} could not be verified.',
    'custom_function'   => '{field} could not be verified.',
    'custom_method'     => '{field} could not be verified.',
];


src/PHPValidation.php
<?php
/**
 * PHPValidation.php
 *
 * This file is part of PHPValidation.
 *
 * @author     Muhammet ŞAFAK <info@muhammetsafak.com.tr>
 * @copyright  Copyright © 2021 PHPValidation
 * @license    http://www.gnu.org/licenses/gpl-3.0.txt  GNU GPL 3.0
 * @version    1.1
 * @link       https://www.muhammetsafak.com.tr
 */

declare(strict_types=1);

namespace PHPValidation;

class PHPValidation
{

    use ValidationMethods;

    protected const FLOAT_PATTERN = '[0-9\.,]+';
    protected const INT_PATTERN = '[0-9]+';
    protected const ALPHA_PATTERN = '[\p{L}]+';
    protected const ALPHA_NUMERIC_PATTERN = '[\p{L}0-9]+';

    protected const CREDIT_CARD_PATTERNS = [
        'amex'          => '(3[47]\d{13})',
        'visa'          => '(4\d{12}(?:\d{3})?)',
        'mastercard'    => '(5[1-5]\d{14})',
        'maestro'       => '((?:5020|5038|6304|6579|6761)\d{12}(?:\d\d)?)',
        'jcb'           => '(35[2-8][89]\d\d\d{10})',
        'solo'          => '((?:6334|6767)\d{12}(?:\d\d)?\d?)',
        'switch'        => '(?:(?:(?:4903|4905|4911|4936|6333|6759)\d{12})|(?:(?:564182|633110)\d{10})(\d\d)?\d?)',
    ];

    private string $version = '1.1';

    protected string $repeatSuffix = '_repeat';

    protected array $patterns = [
        'uri' => '[A-Za-z0-9-\/_?&=]+',
        'slug' => '[-a-z0-9_-]',
        'url' => '[A-Za-z0-9-:.\/_?&=#]+',
        'alpha' => '[\p{L}]+',
        'words' => '[\p{L}\s]+',
        'alphanum' => '[\p{L}0-9]+',
        'int' => '[0-9]+',
        'float' => '[0-9\.,]+',
        'tel' => '[0-9+\s()-]+',
        'text' => '[\p{L}0-9\s-.,;:!"%&()?+\'°#\/@]+',
        'file' => '[\p{L}\s0-9-_!%&()=\[\]#@,.;+]+\.[A-Za-z0-9]{2,4}',
        'folder' => '[\p{L}\s0-9-_!%&()=\[\]#@,.;+]+',
        'address' => '[\p{L}0-9\s.,()°-]+',
        'date_dmy' => '[0-9]{1,2}\-[0-9]{1,2}\-[0-9]{4}',
        'date_ymd' => '[0-9]{4}\-[0-9]{1,2}\-[0-9]{1,2}',
        'email' => '[a-zA-Z0-9_.-]+@[a-zA-Z0-9-]+.[a-zA-Z0-9-.]+[.]+[a-z-A-Z]'
    ];

    protected string $localeFile = 'en.php';

    protected array $internalValidLanguages = [
        'en', 'tr',
    ];

    protected array $locale = [
        'labels'            => [],
        'integer'           => '{field} must be an integer.',
        'float'             => '{field} must be an float.',
        'numeric'           => '{field} must be an numeric.',
        'string'            => '{field} must be an string.',
        'boolean'           => '{field} must be an boolean',
        'array'             => '{field} must be an Array.',
        'mail'              => '{field} must be an E-Mail address.',
        'mailHost'          => '{field} the email must be a {2} mail.',
        'url'               => '{field} must be an URL address.',
        'urlHost'           => 'The host of the {field} url must be {2}.',
        'empty'             => '{field} must be empty.',
        'required'          => '{field} cannot be left blank.',
        'min'               => '{field} must be greater than or equal to {2}.',
        'max'               => '{field} must be no more than {2}.',
        'length'            => 'The {field} length range must be {2}.',
        'regex'             => '{field} must match the {2} pattern.',
        'date'              => '{field} must be a date.',
        'dateFormat'        => '{field} must be a correct date format.',
        'ip'                => '{field} must be the IP Address.',
        'ipv4'              => '{field} must be the IPv4 Address.',
        'ipv6'              => '{field} must be the IPv6 Address.',
        'repeat'            => '{field} must be the same as {field1}',
        'equals'            => '{field} can only be {2}.',
        'startWith'         => '{field} must start with "{2}".',
        'endWith'           => '{field} must end with "{2}".',
        'in'                => '{field} must contain {2}.',
        'notIn'             => '{field} must not contain {2}.',
        'alpha'             => '{field} must contain only alpha characters.',
        'alphaNum'          => '{field} can only be alphanumeric.',
        'creditCard'        => '{field} must be a credit card number.',
        'only'              => 'The {field} value is not valid.',
        'strictOnly'        => 'The {field} value is not valid.',
        'custom_closure'    => '{field} could not be verified.',
        'custom_function'   => '{field} could not be verified.',
        'custom_method'     => '{field} could not be verified.',
    ];

    protected array $rules = [
        'integer',
        'float',
        'numeric',
        'string',
        'boolean',
        'array',
        'mail',
        'mailHost',
        'url',
        'urlHost',
        'empty',
        'required',
        'min',
        'max',
        'length',
        'regex',
        'date',
        'dateFormat',
        'ip',
        'ipv4',
        'ipv6',
        'repeat',
        'equals',
        'startWith',
        'endWith',
        'in',
        'notIn',
        'alpha',
        'alphaNum',
        'creditCard',
        'only',
        'strictOnly',
    ];

    protected array $error = [];

    protected array $userError = [];

    protected array $data = [];

    protected array $rule = [];

    protected array $optional = [];

    protected array $customRules = [];

    public function __construct(array $data = [])
    {
        $this->data = $data;
    }

    /**
     * Imports translations from the localization file.
     *
     * @throws \Exception
     * @param string|null $localePath
     * @return $this
     */
    public function locale(?string $localePath = null): self
    {
        if($localePath === null){
            $path = \rtrim(__DIR__, \DIRECTORY_SEPARATOR) . \DIRECTORY_SEPARATOR . 'lang' . \DIRECTORY_SEPARATOR . 'en.php';
        }elseif(\in_array($localePath, $this->internalValidLanguages, true)){
            $path = \rtrim(__DIR__, \DIRECTORY_SEPARATOR) . \DIRECTORY_SEPARATOR . 'lang' . \DIRECTORY_SEPARATOR . $localePath . '.php';
        }else{
            $path = $localePath;
        }
        if(!\is_file($path)){
            throw new \Exception('PHPValidation: The localization file could not be found.');
        }
        $this->localeFile = $path;
        $locale = $this->locateLoad();
        if(\is_array($locale)){
            $this->localeArray($locale);
        }
        return $this;
    }

    /**
     * Merges an array with a locale (language) array.
     *
     * @param array $localeArray
     * @return $this
     */
    public function localeArray(array $localeArray): self
    {
        $this->locale = \array_merge($this->locale, $localeArray);
        return $this;
    }

    /**
     * Returns the version of the library.
     *
     * @return string
     */
    public function version(): string
    {
        return $this->version;
    }

    /**
     * Returns a copy of the library, loading the specified data directory.
     *
     * @param array $data
     * @return $this
     */
    public function withData(array $data = []): self
    {
        $clone = clone $this;
        $clone->data = \array_merge($clone->error, $data);
        return $clone;
    }

    /**
     * Defines a named regular expression pattern.
     *
     * @param string $name
     * @param string $pattern
     * @return $this
     */
    public function pattern(string $name, string $pattern): self
    {
        $name = \strtolower($name);
        $this->patterns[$name] = $pattern;
        return $this;
    }

    /**
     * Adds a rule.
     *
     * @param string|array $rule <p>You can define multiple rules at once as a array or as a string separated by "|".</p>
     * @param string|array $dataKey <p>Defines the keys of the data to which the defined rules will be applied. If you want to add more than one, you can specify it as an array or as a string separated by "|".</p>
     * @return self
     */
    public function rule($rule, $dataKey): self
    {
        if(\is_string($rule)){
            if(\strpos($rule, '|')){
                $rule = \explode('|', $rule);
            }else{
                $rule = [$rule];
            }
        }
        if(!\is_array($rule)){
            throw new \InvalidArgumentException('Rules can be specified as a string or array.');
        }
        $dataKey = $this->prepareData($dataKey);
        $rules = [];
        foreach($rule as $row){
            if($row == 'optional'){
                $this->optional = \array_merge($this->optional, $dataKey);
            }
            $ruleClear = \preg_replace('/\((.*)\)/u', '', $row);
            if(\in_array($ruleClear, $this->rules, true) !== FALSE){
                $rules[] = $row;
            }else{
                $this->userError[] = '"'.$row.'" validation failed because the rule you specified could not be found.';
            }
        }
        $this->rule[] = [
            "rule" => $rules,
            "dataKey" => $dataKey,
        ];

        return $this;
    }

    /**
     * It allows you to write your own validation method with a callable function.
     *
     * @param \Closure $closure
     * @param string|array $dataKey
     * @param string $errorMsg
     * @return $this
     */
    public function ruleClosure(\Closure $closure, $dataKey, string $errorMsg = ''): self
    {
        $this->customRules['closure'][] = [
            'closure'   => $closure,
            'dataKey'    => $this->prepareData($dataKey),
            'errorMsg'  => $errorMsg
        ];
        return $this;
    }

    /**
     * Defines a rule using a predefined function.
     *
     * @param string $function
     * @param string|array $dataKey
     * @param string $errorMsg
     * @return $this
     */
    public function ruleFunction(string $function, $dataKey, string $errorMsg = ''): self
    {
        $this->customRules['function'][] = [
            'function'  => $function,
            'dataKey'    => $this->prepareData($dataKey),
            'errorMsg'  => $errorMsg
        ];
        return $this;
    }

    /**
     * Defines a class or object's method and rule.
     *
     * @param $classOrObject
     * @param string $method
     * @param string|array $dataKey
     * @param string $errorMsg
     * @return $this
     */
    public function ruleMethod($classOrObject, string $method, $dataKey, string $errorMsg = ''): self
    {
        if(!\is_object($classOrObject) && !\class_exists($classOrObject)){
            throw new \InvalidArgumentException('A class or object must be specified.');
        }
        $this->customRules['method'][] = [
            'class'     => $classOrObject,
            'method'    => $method,
            'dataKey'    => $this->prepareData($dataKey),
            'errorMsg'  => $errorMsg
        ];
        return $this;
    }

    /**
     * Validates data within specified rules.
     *
     * @return bool
     */
    public function validation(): bool
    {
        $this->error = [];
        $errors = $this->userError;
        $this->userError = [];
        foreach ($this->rule as $rule) {
            foreach ($rule['rule'] as $rule_row) {
                foreach ($rule['dataKey'] as $data_row) {
                    if(!isset($this->data[$data_row]) && \in_array($data_row, $this->optional, true)){
                        continue;
                    }
                    $this->ruleExecutive($rule_row, $data_row);
                }
            }
        }
        if(!empty($this->customRules)){
            $this->customRulesExecutive();
            $this->customRules = [];
        }
        $this->repeatDataValidation();
        $errors = \array_merge($this->error, $errors);
        if (\count($errors) > 0) {
            $return = false;
        } else {
            $return = true;
        }
        $this->rule = [];
        $this->customRules = [];
        return $return;
    }

    private function repeatDataValidation()
    {
        $repeatSuffixLen = \mb_strlen($this->repeatSuffix);
        foreach($this->data as $key => $value){
            if(\mb_substr($key, -$repeatSuffixLen) === $this->repeatSuffix){
                $origDataKey = \mb_substr($key, 0, -$repeatSuffixLen);
                $origData = $this->data[$origDataKey] ?? null;
                if($this->repeat($origData, $key) === FALSE){
                    $this->error[] = $this->__r($origDataKey, 'repeat', ['field' => $origDataKey, 'field1' => $key, $key]);
                }
            }
        }
    }

    /**
     * Returns the value of a data.
     *
     * @param string $key
     * @return mixed|null
     */
    public function data(string $key)
    {
        return $this->data[$key] ?? null;
    }

    /**
     * It adds an external error so that the validation fails.
     *
     * @param string $error
     * @param array $context
     * @return $this
     */
    public function error(string $error, array $context = []): self
    {
        $this->userError[] = $this->interpolate($error, $context);
        return $this;
    }

    /**
     * Returns the errors that occurred as an array. An empty array is returned if there are no errors.
     *
     * @return array
     */
    public function errors(): array
    {
        return \array_merge($this->error, $this->userError);
    }

    public function labels(array $labels): self
    {
        $this->locale['labels'] = \array_merge($this->locale['labels'], $labels);
        return $this;
    }

    private function __r($field, $key, array $context = []): string
    {
        $lang = $this->locale[$field][$key] ?? ($this->locale[$key] ?? $key);
        return $this->interpolate($lang, $context);
    }

    private function interpolate(string $message, array $context = []): string
    {
        $replace = array();
        $i = 0;
        foreach ($context as $key => $val) {
            if (!\is_array($val) && (!\is_object($val) || \method_exists($val, '__toString'))) {
                if((\is_string($val) || \is_int($val) && \substr((string)$key, 0, 5) == 'field')){
                    $val = $this->locale['labels'][$val] ?? $val;
                }
                $replace['{' . $key . '}'] = $val;
                $replace['{' . $i . '}'] = $val;
                $i++;
            }
        }

        return \strtr($message, $replace);
    }

    private function locateLoad()
    {
        return require_once $this->localeFile;
    }

    protected function ruleExecutive(string $rule, string $dataKey): void
    {
        $prepare = $this->prepareArguments($rule, $dataKey);
        $method = $prepare['method'];
        $arguments = $prepare['arguments'];
        if($this->{$method}(...$arguments) === FALSE){
            $field = ['field' => $dataKey];
            $this->error[] = $this->__r($dataKey, $method, \array_merge($field, $arguments));
        }
    }

    protected function customRulesExecutive(): void
    {
        $closureRules = $this->customRules['closure'] ?? [];
        foreach($closureRules as $rule){
            foreach($rule['dataKey'] as $item){
                if(!isset($this->data[$item]) && \in_array($item, $this->optional, true)){
                    continue;
                }
                $data = $this->data[$item] ?? null;
                if(\call_user_func_array($rule['closure'], [$data]) === FALSE){
                    $error = \trim($rule['errorMsg']);
                    $this->error[] = ($error !== '' ? $this->__r($item, $error, ['field' => $item]) : $this->__r($item,'custom_closure', ['field' => $item]));
                }
            }
        }

        $functionRules = $this->customRules['function'] ?? [];
        foreach($functionRules as $rule){
            foreach($rule['dataKey'] as $item){
                if(!isset($this->data[$item]) && \in_array($item, $this->optional, true)){
                    continue;
                }
                $prepare = $this->prepareArguments($rule['function'], $item);
                if(\call_user_func_array($prepare['method'], $prepare['arguments']) === FALSE){
                    $error = \trim($rule['errorMsg']);
                    $this->error[] = ($error !== '' ? $this->__r($item, $error, ['field' => $item]) : $this->__r($item,'custom_function', ['field' => $item]));
                }
            }
        }

        $methodRules = $this->customRules['method'] ?? [];
        foreach($methodRules as $rule){
            $object = $rule['class'];
            if(!\is_object($rule['class'])){
                $object = new $rule['class']();
            }
            foreach($rule['dataKey'] as $item){
                if(!isset($this->data[$item]) && \in_array($item, $this->optional, true)){
                    continue;
                }
                $data = $this->data[$item] ?? null;
                $prepare = $this->prepareArguments($rule['method'], $item);
                if(\call_user_func_array([$object, $prepare['method']], $prepare['arguments']) === FALSE){
                    $error = \trim($rule['errorMsg']);
                    $this->error[] = ($error !== '' ? $this->__r($item, $error, ['field' => $item]) : $this->__r($item,'custom_method', ['field' => $item]));
                }
            }
        }
    }

    private function prepareArguments(string $rule, $dataKey): array
    {
        $arguments = [($this->data[$dataKey] ?? null)];
        \preg_match("/\((.*)\)/u", $rule, $params);
        if(isset($params[0])){
            $method = \preg_replace("/\((.*)\)/u", '', $rule);
            $arguments = \array_merge_recursive($arguments, \explode(',', \trim($params[0], '()')));
        }else{
            $method = $rule;
        }
        return [
            'method'        => $method,
            'arguments'     => $arguments
        ];
    }

    /**
     * @param string|array $dataKey
     */
    protected function prepareData($dataKey): array
    {
        $keys = $dataKey;
        if(\is_string($dataKey)){
            if(\strpos($dataKey, '|')){
                $keys = \explode('|', $dataKey);
            }else{
                $keys = [$dataKey];
            }
        }
        if(!\is_array($keys)){
            throw new \InvalidArgumentException('The keys of the data to be validated must be specified as a string or array.');
        }
        return $keys;
    }

}
